### PR TITLE
Infrastructure/2556 google cloud platform terraform module

### PR DESCRIPTION
**Pull request by bilgeAiry**
- The module for gcp-gke is created. 
- The kubeconfig file is currently not possible to update.

### CODE REVIEW

**Comment by bilgeAiry** [comment]
deleted some scripts in install directory from 3 months ago.. 😨 what is wrong with me! Thanks @ljupcovangelski they are back now

File: infrastructure/terraform/modules/gcp-gke/main.tf [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  region  = var.region
+}
+
+# VPC
+resource "google_compute_network" "vpc" {
+  name                    = var.vpc_name
+  auto_create_subnetworks = "false"
+}
+
+# Subnet # private_ip_google_access = false by default
```
Comment (ljupcovangelski) [inline infrastructure/terraform/modules/gcp-gke/main.tf:12]: We don't need these comments

File: infrastructure/terraform/modules/gcp-gke/main.tf [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+# GKE cluster
+resource "google_container_cluster" "primary" {
+  name     = "${var.project_id}-gke"
+  location = var.zone
+  
+  # We can't create a cluster with no node pool defined, but we want to only use
+  # separately managed node pools. So, we create the smallest possible default
+  # node pool and immediately delete it.
```
Comment (ljupcovangelski) [inline infrastructure/terraform/modules/gcp-gke/main.tf:27]: Also these comments we can remove I think.

File: infrastructure/terraform/modules/gcp-gke/main.tf [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+  project_id   = var.project_id
+  location     = google_container_cluster.primary.location
+  cluster_name = google_container_cluster.primary.name
+}
+
+resource "local_file" "kubeconfig" {
+  content  = module.gke_auth.kubeconfig_raw
+  filename = var.kubeconfig_output_path
+}
```
Comment (ljupcovangelski) [inline infrastructure/terraform/modules/gcp-gke/main.tf:76]: The content of the file is not with the final credentials. Perhaps run the `module.gke_auth` after `resource.google_container_cluster`, with the `depends_on` argument.

File: infrastructure/terraform/modules/gcp-gke/outputs.tf [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+}
+output "kubeconfig_raw" {
+  sensitive = true
+  value = module.gke_auth.kubeconfig_raw
+}
+output "kubernetes_endpoint" {
+  sensitive = true
+  value     = module.gke_auth.host
+}
+output "ca_certificate" {
```
Comment (ljupcovangelski) [inline infrastructure/terraform/modules/gcp-gke/outputs.tf:29]: Perhaps we don't need this output? As we already have the kubeconfig file.

File: infrastructure/terraform/modules/gcp-gke/variables.tf [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  description = "The zone defined in gcloud config is us-central1-a"
+}
+variable "gke_num_nodes" {
+  default     = 2
+  description = "number of gke nodes"
+}
+variable "vpc_name" {
+  default = "airy-core-vpc"
+}
+variable "private_subnets" {
```
Comment (ljupcovangelski) [inline infrastructure/terraform/modules/gcp-gke/variables.tf:23]: The subnets are not used, so we can also remove the variables.

File: infrastructure/terraform/modules/gcp-gke/variables.tf [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+}
+variable "public_subnets" {
+  type        = list(string)
+  default = ["10.0.4.0/24", "10.0.5.0/24", "10.0.6.0/24"]
+}
+variable "network" {
+  description = "The VPC network created to host the cluster in"
+  default     = "gke-network"
+}
+variable "subnetwork" {
```
Comment (ljupcovangelski) [inline infrastructure/terraform/modules/gcp-gke/variables.tf:35]: This variable is also not used, so not needed I guess. Same for the remaining 5 variables I think?

**Review by ljupcovangelski — COMMENTED** [review]
Looks good and it is working properly :+1: 
Left some minor comments.

File: infrastructure/terraform/modules/gcp-gke/variables.tf (hunk 5 above)
Comment (bilgeAiry) [reply infrastructure/terraform/modules/gcp-gke/variables.tf:23]: 👍🏼

File: infrastructure/terraform/modules/gcp-gke/variables.tf (hunk 6 above)
Comment (bilgeAiry) [reply infrastructure/terraform/modules/gcp-gke/variables.tf:35]: 👍🏼

File: infrastructure/terraform/modules/gcp-gke/main.tf (hunk 1 above)
Comment (bilgeAiry) [reply infrastructure/terraform/modules/gcp-gke/main.tf:12]: 👍🏼

File: infrastructure/terraform/modules/gcp-gke/main.tf (hunk 2 above)
Comment (bilgeAiry) [reply infrastructure/terraform/modules/gcp-gke/main.tf:27]: 👍🏼

File: infrastructure/terraform/modules/gcp-gke/outputs.tf (hunk 4 above)
Comment (bilgeAiry) [reply infrastructure/terraform/modules/gcp-gke/outputs.tf:29]: 👍🏼

File: infrastructure/terraform/modules/gcp-gke/main.tf [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+# GKE cluster
+resource "google_container_cluster" "gke_core" {
+  name     = "${var.project_id}-gke"
+  location = var.zone
+  
+  remove_default_node_pool = true
+  initial_node_count       = 1
+
+  network    = google_compute_network.vpc.name
+  subnetwork = google_compute_subnetwork.subnet.name
```
Comment (ljupcovangelski) [inline infrastructure/terraform/modules/gcp-gke/main.tf:29]: This parameter is optional I think, so try without it. That way we can also remove the subnetwork resource and the public ip range.

**Comment by bilgeAiry** [comment]
`https://stackoverflow.com/questions/59004600/terraform-create-single-node-gke-cluster` an example for zonal nodes and regional nodes issue in GKE

Now the GKE-VPC is in the defaults (in global location setting) and with it the cluster has 35 subnets
The two googleapis is used in oauth scopes for monitoring and writing the kube.conf file
Extra variables deleted after the review.

File: infrastructure/terraform/modules/gcp-gke/main.tf (hunk 7 above)
Comment (bilgeAiry) [reply infrastructure/terraform/modules/gcp-gke/main.tf]: We can't create a GCP cluster with no node pool defined, but we want to only use separately managed node pools. So we create the smallest possible default node pool and immediately delete it.

Comment (bilgeAiry) [reply infrastructure/terraform/modules/gcp-gke/main.tf]: And the subnetwork there deleted

File: infrastructure/terraform/modules/gcp-gke/main.tf [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+  name       = "${google_container_cluster.gke_core.name}-node-pool"
+  location   = var.region
+  cluster    = google_container_cluster.gke_core.name
+  node_count = var.gke_num_nodes
+
+  node_config {
+    preemptible  = true
+    machine_type = var.gke_instance_type
+
+    oauth_scopes = [
```
Comment (bilgeAiry) [inline infrastructure/terraform/modules/gcp-gke/main.tf:30]: We need here for kube.cof file

File: infrastructure/terraform/modules/gcp-gke/main.tf [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+      "https://www.googleapis.com/auth/logging.write",
+      "https://www.googleapis.com/auth/monitoring",
+    ]
+
```
Comment (bilgeAiry) [inline infrastructure/terraform/modules/gcp-gke/main.tf:34]: We need there two for kube.conf file

File: infrastructure/terraform/modules/gcp-gke/main.tf [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+resource "google_container_node_pool" "gke_core_nodes" {
+  name       = "${google_container_cluster.gke_core.name}-node-pool"
+  location   = var.region
+  cluster    = google_container_cluster.gke_core.name
+  node_count = var.gke_num_nodes
+
+  node_config {
+    preemptible  = true
```
Comment (bilgeAiry) [inline infrastructure/terraform/modules/gcp-gke/main.tf:27]: well having it preemptible is more budget..

**Review by bilgeAiry — COMMENTED** [review]
🤞🏼

**Review by ljupcovangelski — APPROVED** [review]
Works great :tada:

**Pull request merged**